src/lib/sql/sql_pipeline.hpp

```diff
@@ -0,0 +1,87 @@
+#pragma once
+
+#include <string>
+
+#include "SQLParserResult.h"
+#include "concurrency/transaction_context.hpp"
+#include "logical_query_plan/abstract_lqp_node.hpp"
+#include "sql/sql_query_plan.hpp"
+#include "storage/table.hpp"
+
+namespace opossum {
+
+/**
+ * This is the unified interface to handle SQL queries and related operations.
+ *
+ * The basic idea of the SQLPipeline is that it represents the flow from the basic SQL string to the result table with
+ * all intermediate steps. These intermediate steps call the previous step that is required. The intermediate results
+ * are all cached so calling a method twice will return the already existing value.
+ *
+ * The SQLPipeline holds all results and only hands them out as const references. If the SQLPipeline goes out of scope
+ * while the results are still needed, the result references are invalid (except maybe the result_table).
+ *
+ * E.g: calling sql_pipeline.get_result_table() will result in the following "call stack"
+ * get_result_table -> get_tasks -> get_query_plan -> get_optimized_logical_plan -> get_parsed_sql
+ */
+class SQLPipeline : public Noncopyable {
+ public:
+  // An explicitly passed TransactionContext will result in no auto-commit and LQP validation.
+  SQLPipeline(const std::string& sql, std::shared_ptr<TransactionContext> transaction_context);
+
+  // If use_mvcc is true, a Transaction context is created with auto-commit and LQP validation.
+  // If false, no validation or transaction management is done.
+  explicit SQLPipeline(const std::string& sql, bool use_mvcc = true);
+
+  const std::string sql_string() { return _sql_string; }
+
+  // Returns the parsed SQL string.
+  const hsql::SQLParserResult& get_parsed_sql();
+
+  // Returns all unoptimized LQP roots.
+  const std::vector<std::shared_ptr<AbstractLQPNode>>& get_unoptimized_logical_plan();
+
+  // Returns all optimized LQP roots.
+  const std::vector<std::shared_ptr<AbstractLQPNode>>& get_optimized_logical_plan();
+
+  // For now, this always uses the optimized LQP.
+  const SQLQueryPlan& get_query_plan();
+
+  // Returns all tasks tht need to be executed for this query.
+  const std::vector<std::shared_ptr<OperatorTask>>& get_tasks();
+
+  // Executes all tasks, waits for them to finish, and returns the resulting table.
+  const std::shared_ptr<const Table>& get_result_table();
+
+  // Returns the TransactionContext that was either passed to or created by the SQLPipeline.
+  // This can be a nullptr if no transaction management is wanted.
+  const std::shared_ptr<TransactionContext>& transaction_context();
+
+  float parse_time_seconds();
```

Comment:
sorry, should have spotted this earlier, but why use a float at all when there is a dedicated duration type in the stdlib. feel free to create a follow up issue, it's probably not that important.

Reply:
You are probably right, but we only use this time information in the console, so I would keep this simple for now. If we want to use the time in other scenarios, e.g. network interface and we need more sophisticated operations on time than just printing it, we can definitely change it then.

Reply:
I hate std::chrono (imo all that c++ would have needed was a type `nanos_t` and a function `current_nanos()`, but that's just me....) with a passion and feel bad for pushing for it, but IME if the "we can definitely change it then." isn't doc'ed somehow (either Issue or // TODO) instead of fixing it people will just work around it - and then probably switch back to double or commit other atrocities.

Reply:
Ok, I'll try to include it now then. 
Just to make sure, this is the way you want to have it?

```cpp
std::chrono::duration<float> execution_time_seconds();
```
and
```cpp
_execution_time_sec = std::chrono::duration<float>(done - started);
```
Or is there a more concise way of returning a duration?
So the time stored is still in seconds but we can do further time operations on it if needed.

Reply:
There are typedefs for microseconds and nanoseconds
http://en.cppreference.com/w/cpp/chrono/duration